crates/lib3h/src/engine/real_engine.rs

```diff
@@ -554,6 +503,58 @@ impl<T: Transport, D: Dht, SecBuf: Buffer, Crypto: CryptoSystem> RealEngine<T, D
         Ok(output)
     }
 
+    fn serve_DirectMessage(
+        &mut self,
+        msg: DirectMessageData,
+        is_response: bool,
+    ) -> Lib3hServerProtocol {
+        // Check if space is joined by sender
+        let maybe_space = self.get_space_or_fail(
+            &msg.space_address,
+            &msg.from_agent_id,
+            &msg.request_id,
+            None,
+        );
+        // Return failure if not
+        if let Err(failure_msg) = maybe_space {
+            return failure_msg;
+        }
+        let space_gateway = maybe_space.unwrap();
+        // Prepare response
+        let mut response = GenericResultData {
+            request_id: msg.request_id.clone(),
+            space_address: msg.space_address.clone(),
+            to_agent_id: msg.from_agent_id.clone(),
+            result_info: vec![],
+        };
+        // Check if messaging self
+        let this_peer = space_gateway.this_peer();
+        if this_peer.peer_address.as_bytes() == msg.to_agent_id.as_slice() {
+            response.result_info = "Messaging self".as_bytes().to_vec();
+            return Lib3hServerProtocol::FailureResult(response);
+        }
+        // Change into P2pProtocol
+        let net_msg = if is_response {
+            P2pProtocol::DirectMessageResult(msg.clone())
+        } else {
+            P2pProtocol::DirectMessage(msg.clone())
+        };
+        // Serialize payload
+        let mut payload = Vec::new();
+        net_msg
+            .serialize(&mut Serializer::new(&mut payload))
+            .unwrap();
```

Comment:
If this serialization errors happen easily I would suggest returning it back as a `FailureResult`. I don't know when that would actually occur though.

Reply:
Agree. I believe serialization like this can't ever fail so no need for it. I believe it returns a Result because its a trait and that other implemetations could fail.